File: LeetCode/MedianOfTwoSortedArrays.py
class Solution:
    # @return a float
    def findMedianSortedArrays(self, A, B):
         n = len(A)
         m = len(B)
         ans = self.get(A, 0, n, B, 0, m, (m+n)/2+1)
         if (m + n) % 2 == 1:
             return ans
         else:
             return (ans + self.get(A, 0, n, B, 0, m, (m+n)/2)) / 2.0
    def get(self, A, a, n, B, b, m, k):
        if n <= 0:
            return B[b+k-1]
        if m <= 0:
            return A[a+k-1]
        if k <= 1:
            return min(A[a], B[b])
        if A[a+n/2] <= B[b+m/2]:
            if n/2 + 1 + m/2 >= k:
                return self.get(A, a, n, B, b, m/2, k)
            else:
                return self.get(A, a + n/2 + 1, n - (n/2 + 1), B, b, m, k - (n/2 + 1))
        else:
            if m/2 + 1 + n/2 >= k :
                return self.get(A, a, n/2, B, b, m, k)
            else:
                return self.get(A, a, n, B, b + m/2 + 1, m - (m/2 + 1), k - (m/2 + 1))

File: LeetCode/MedianOfTwoSortedArrays.cc
class Solution {
public:
    double findMedianSortedArrays(int A[], int m, int B[], int n) {
        // Note: The Solution object is instantiated only once and is reused by each test case.
        int cutA = m/2, cutB = n/2;
        if(m <= 2 || n <= 2) {
            vector<int> v;
            for(int i = 0; i < m; i++)
                v.push_back(A[i]);
            for(int i = 0; i < n; i++)
                v.push_back(B[i]);
            sort(v.begin(), v.end());
            if((m + n) % 2)
                return v[(m+n)/2];
            else
                return double(v[(m+n)/2] + v[(m+n)/2-1])/2;
        }
            
        if(m % 2 == 0 && n % 2 == 0)
            cutA--;
        if(A[cutA] < B[cutB]) {
            int discard = min(cutA, n - 1 - cutB);
            return findMedianSortedArrays(A + discard, m - discard, B, n - discard);
        } else if(A[cutA] > B[cutB]) {
            int discard = min(m - 1 - cutA, cutB);
            return findMedianSortedArrays(A, m - discard, B + discard, n - discard);
        } else {
            return A[cutA];
        }
    }
};